generator/genutils/utils.go

```diff
@@ -0,0 +1,185 @@
+				unionMemberProperty := strcase.ToLowerCamel(fieldName)
+				if _, foundUnionMember := schema.Properties[unionMemberProperty]; !foundUnionMember {
+					// The discriminator enum contains values that do not correspond to any enum field.
+					// So so just skip the oneOf creation here since this union definition seems invalid.
+					continue discriminatorLoop
+				}
+				for _, toSkip := range fieldsToSkip {
+					if toSkip == fieldName {
+						continue enumValueLoop
+					}
+				}
+				newEnumValues = append(newEnumValues, enumValue)
+				oneOf = append(oneOf, apiext.JSONSchemaProps{
+					Required: []string{unionMemberProperty},
+				})
+			}
+			schema.OneOf = oneOf
+			discriminatorProp.Enum = newEnumValues
+
+			if removeDiscriminators {
+				delete(schema.Properties, discriminatorPropertyName)
+			} else {
+				schema.Properties[discriminatorPropertyName] = discriminatorProp
+			}
+			for _, toSkip := range fieldsToSkip {
+				delete(schema.Properties, strcase.ToLowerCamel(toSkip))
+			}
+		}
+		return
+	}
+
+	EditJSONSchema(jsonSchema, mainVisitor)
+}
+
+var (
+	// UnionMarker is the definition of the union marker, as defined in https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#proposal
+	UnionMarker = markers.Must(markers.MakeDefinition("union", markers.DescribesType, struct{}{}))
+	// UnionDiscriminatorMarker is the definition of the union discriminator marker, as defined in https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#proposal
+	UnionDiscriminatorMarker = markers.Must(markers.MakeDefinition("unionDiscriminator", markers.DescribesField, struct{}{}))
+)
+
+// RegisterUnionMarkers registers the `union` and `unionDiscriminator` markers
+func RegisterUnionMarkers(into *markers.Registry) error {
+	if err := markers.RegisterAll(into, UnionMarker, UnionDiscriminatorMarker); err != nil {
+		return err
+	}
+	into.AddHelp(UnionMarker,
+		markers.SimpleHelp("Devfile", "indicates that a given Struct type is a K8S union, and its fields (apart from the discriminator) are mutually exclusive. K8S unions are described here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#proposal"))
+	into.AddHelp(UnionDiscriminatorMarker,
+		markers.SimpleHelp("Devfile", "indicates that a given field of an union Struct type is the union discriminator. K8S unions are described here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#proposal"))
+	return nil
+}
+
+// WriteFormattedSourceFile creates a Go source file in a given package, dumps to it the content provided by the `writeContents` function
+// and formats the result through go/fmt.
+// If formatting cannot be applied (due to some syntax error probably), it returns an error.
+func WriteFormattedSourceFile(filename string, ctx *genall.GenerationContext, root *loader.Package, writeContents func(*bytes.Buffer)) error {
```

Comment:
I wonder it it makes sense to split logic of utils in different files, like marker.go, json_schema.go, patch_strategy.go, file.go like the following https://github.com/devfile/api/commit/1878c78ead018a4ab8444a042237c67f9a35156e

Reply:
I can do it, if you think it would bring more clarity.

Reply:
Done in commit https://github.com/devfile/api/pull/133/commits/7a6fc780cf9a34af3a95f1184aca4fd6b2d8fd43